Proc macro for css

WorldSEnder opened the pull request:
Implements a proc macro for writing css, resulting in an expression of type `Sheet`. The following parts are still missing or up for review already:

- [x] As part of writing this test case https://github.com/futursolo/stylist-rs/compare/master...WorldSEnder:stylist?expand=1#diff-260b6dc657a6ef35ee73a531d3e7f5a8fd197b4ca04663d48276179cf278ac10, issue #16 came up, ~so this also contains a fix for that.~ the resolution has been benched
- [x] ~Additionally fixes failing compilation on master when `--target wasm32-unknown-unknown` is active. I suppose at some point I can rebase against a fix on master~ master has been fixed
- [x] Documentation of the proc macro itself
- [x] Address review comments
- [ ] Round two of reviews

futursolo commented:
Thank you for another PR!

However, I'd like to ask a question:

What's the deciding factor of picking this syntax over string literals?

(I feel harsh to bring this question up after you are thousands of lines in.
But I have tried to avoid this situation by asking the same question in #8 a couple days ago.)

That is:

```rust
let sheet = css! {
    background: red;
};
```

over:

```rust
let sheet = css!(r#"background: red;"#);
```

The reason why I am asking is because that we will be able to reuse the parser if a string literal is used.

For a short period of time, We've got #1 and #12 (and #16 if selector parsing is needed) that needs changes on the parser.
I can foresee that maintaining the parser will certainly take a good chunk of time in the time that I will be maintaining this project. Therefore, I want to avoid the fate of having to maintain two parsers (if possible).

If a string literal is used, the procedural macro will become as simple as something like the following:

(BTW, I did test my code this time.)

```rust
use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;

use litrs::StringLit;
use proc_macro2::Literal;
use std::convert::TryFrom;
use stylist_core::ast::*;

use quote::{quote, quote_spanned};

fn attr_to_token_stream(attr: &StyleAttribute) -> TokenStream2 {
    let key_token = Literal::string(&attr.key);
    let value_token = Literal::string(&attr.value);

    quote! {
        ::stylist::ast::StyleAttribute {
            key: #key_token.to_string(),
            value: #value_token.to_string(),
        }
    }
}

fn block_to_token_stream(block: &Block) -> TokenStream2 {
    let cond = if let Some(ref m) = block.condition {
        let s = Literal::string(m);
        quote! { Some(#s.to_string()) }
    } else {
        quote! { None }
    };

    let mut attrs = TokenStream2::new();

    for attr in block.style_attributes.iter() {
        attrs.extend(attr_to_token_stream(attr));
        attrs.extend(quote! { ,});
    }

    quote! {
       ::stylist::ast::ScopeContent::Block( ::stylist::ast::Block {
            condition: #cond,
            style_attributes: vec![#attrs]
        })
    }
}

fn sheet_to_token_stream(sheet: &Sheet) -> TokenStream2 {
    let mut output = TokenStream2::new();

    for scope in sheet.0.iter() {
        match scope {
            ScopeContent::Block(ref m) => {
                output.extend(block_to_token_stream(m));
            }
            _ => panic!(),
        }

        output.extend(quote! { ,});
    }

    quote! {
        ::stylist::ast::Sheet(vec![#output])
    }
}

#[proc_macro]
pub fn css(input: TokenStream) -> TokenStream {
    let mut tokens = input.into_iter();

    let first_token = tokens.next().unwrap();
    let s_literal = StringLit::try_from(first_token).unwrap();
    let sheet: Sheet = s_literal.value().parse().unwrap();

    sheet_to_token_stream(&sheet).into()
}
```

WorldSEnder commented:
The advantage should be, with correct implementation, error highlighting. E.g. forgetting a `;`, misspelling a css attribute name, etc. are all things that are much more easily given feedback on with `quote_spanned` than with the generic string parsing.

Further, as you can see you can include arbitrary expressions of `impl Display` with the `${...}` syntax. Editor completion inside these brackets works as usual. Scope lookup works as usual.
I see it like this: the proc macro syntax is for editor support that you just can't get with string literals. For that one pays the cost of a second parser.

WorldSEnder commented:
Example of what you can't do as well when you always `parse` (assume a fitting `theme` is in the context)

```rust
css! {
  backgroundColor: ${theme.background};
  fontSize: ${theme.fontSize};
}
```

Converting the above to css should take almost no time, as the syntax tree is statically determined, but the values are injected at runtime. I do not see how a comparable approach could work with using parse, as the style string is just not available at compile time.

WorldSEnder commented:
I should also mention, I do value your input on this, you seem level headed and thoughtful :)

But don't cry for the time I sink into this, I would do it either way and at the end of the day, I will have to eat my own dogfood code in [material-yewi](/worldsender/material-yewi) so at least it will be somewhat battle tested.

futursolo commented:
> I should also mention, I do value your input on this, you seem level headed and thoughtful :)

How could I deserve such an admiration. You are a humourous person. (I mean, `yewi`. ;) )

The parser already reports the exact location on errors, when combined with `proc_macro_error`, you get:

<img width="916" alt="image" src="https://user-images.githubusercontent.com/11693215/129200276-9beb6101-ccdf-4cee-98d6-f999d32fe36a.png">

Comparing to Native error:

<img width="834" alt="image" src="https://user-images.githubusercontent.com/11693215/129199959-36c95a85-837c-411a-a6f6-f10bf240b050.png">

Both are not precisely reporting the location, but I think both are optimisable (and acceptable for now).

It's also possible to provide error handling on interpolation values at compile time with format!-style syntax using `quote_spanned`:
```rust
let color = Color::from_rgb(255, 255, 255);  // implements Display

let css = css!(
    r#"
        color: ${color},
    "#,
    color = color, // quote_spanned! {} here
);
```

In this form, editor completion should work as well as if it's outside of a proc macro.

We just need to expand the existing parser to accept interpolation during compile time and reject interpolation during runtime and limit interpolation to only value of attributes and conditions.

Also not related but having a format!-style syntax also has other benefits (does not have to be supported this round):

```rust
let opacity = 0.1;

let css = css!(
    r#"
        /* Limit precision to 0.1 */
        opacity: ${opactity:.1};
    "#,
    opacity = opacity,
);
```
Although, I suspect this would also be possible for non-string literal syntax.

futursolo commented:
Although, I am still slightly leaning towards a format!-like string literal syntax, purely based on it's less effort to maintain.

By the end of the day, I will accept either syntax as I am not the person implementing this PR (unresponsible). :)

futursolo commented:
Since there's no development on this PR for a while and I want to experiment with #23, I have started to implemented a tentative procedural macro in master and have re-purposed `css!` for a different API.

The API that emits a `Sheet` instance has been renamed to `stylist::ast::sheet!`.

I will still accept this PR if you want to implement this as an alternative syntax.

WorldSEnder commented:
I was on vacation for a week, I'll read through what you've done in the mean time, looks promising. I really hope I can switch to this so I can test impacts on code size :+1:

WorldSEnder commented:
@futursolo if you have 10 minutes time, can you give me a quick code review focused on readability, naming and points where to put internal code documentation? I'm going to work on the missing user-facing documentation now, apart from that, I consider the PR to be feature complete.

futursolo reviewed (changes requested):
The current token based parser might be "too complex" (both in amount of code and difficulty to understand) and I would like to see the complexity reduced before merging.

There're a couple ways that this could be done:
1. follow rust naming convention (e.g:`stringify` -> `to_string`)
2. `abort!` on first error and move all error handling into parse stage (by using `proc_macro_error`).
3. Follow `Token` -> `AST(Css*)` -> AST(`stylist::ast::*`) more closely (Parse selectors into selectors in parsing stage other than seeking for commas in output stage)
4. Remove `Output*` and `fold_*` by not doing anything that is not supported by `stylist::ast`.
5. Collect tokens into string literal whenever possible so output stage don't have to consider them.
6. Document what you are trying to in the complex code.

I recently discovered a crate [jss](https://crates.io/crates/jss) and I think their syntax is more elegant and simple to handle (`json::object!` is not even a procedural macro.).

Maybe shifting our syntax a little towards `jss` can help reducing the complexity (use some string literals, as there're cases where css tokens cannot be represented properly (like: `#000000`))?

WorldSEnder commented on Cargo.toml:
```diff
@@ -3,7 +3,7 @@ members = [
     "packages/stylist-core",
     "packages/stylist-macros",
-
+    "packages/stylist-macro-utils",
```
So doc-testing works. Can't export the trait from the procmacro crate, so it isn't accessible to the doctest. If you know a way, I'd be glad to put it back :)

futursolo replied:
You can simply remove the example from doc string and use normal comments.
This trait shouldn't be visible in docs when present `-macros` anyways, and viewing comments and doc strings does not make a difference.

Also this trait is probably over-engineered, if in the actual case a space is always a ' '(actual space).

WorldSEnder replied:
I was hoping that at some point in the future one'd be able to swap it out for an existing method in `itertools` but wasn't able to find one. I think this somewhat over-engineered approach would be fastest to swap out in the future.

WorldSEnder commented on packages/stylist-macros/src/tokentree/css_ident.rs:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    pub fn quote_at_rule(&self) -> LitStr {
+        let formatted = self.stringify().from_case(Case::Camel).to_case(Case::Kebab);
```
Sensible, I was thinking that when one just wants to copy paste over existing code from emotion, it'd be nice to have. You think an error makes sense when non-kebab style is detected?

futursolo replied:
> copy paste over existing code from emotion

You cannot really do that.

Because the Dict / HashMap style of emotion uses comma other than colon and all values are quoted strings.

WorldSEnder commented on packages/stylist-macros/src/tokentree/output/rule.rs:
```diff
@@ -0,0 +1,52 @@
+    pub name: TokenStream,
+    pub prelude: Vec<ComponentValue>,
+    pub contents: Vec<TokenStream>,
+    pub errors: Vec<ParseError>,
```
Ah I was under the impression that issuing diagnostics was nightly only. I will change those to `proc_macro_error::emit_error`.

WorldSEnder replied:
Indeed, diagnostics do change for the worse. In fact, `proc_macro_error::emit_error` seems not to be up to date with `ToTokens` and defaults to marking the whole macro as an error, which is not very helpful, in some cases. Using `syn::Error` leads to much more precise error locations.

I want to emit as many diagnostics as possible to the user if the parser is able to recover somehow.

futursolo replied:
This might be a bug with `proc_macro_error`.

I guess we can keep `syn::Error` for now and I will open an issue with `proc_macro_error`.

futursolo commented on packages/stylist-macros/src/tokentree/parsed.rs:
```diff
@@ -0,0 +1,535 @@
+        let mut prelude = vec![];
+        let mut errors = vec![];
+        if !["media", "supports"].contains(&name.stringify().as_str()) {
+            errors.push(ParseError::new_spanned(
```
I think we can leave this as error as there're more works to support parsing arbitrary at rules with tokens.

WorldSEnder commented on packages/stylist-macros/src/tokentree/output/mod.rs:
```diff
@@ -0,0 +1,44 @@
+pub use str_frag::{fragment_coalesce, fragment_spacing, OutputFragment};
+
+/// Reify a structure into an expression of a specific type.
+pub(crate) trait Reify {
```
I can see why you'd want to use `Lazy` by default in the `css!` and `global_style!` macros, but overall, I think it's also overcomplicating things.

I'd be in favor of mentioning that in the documentation and letting the user wrap everything inside the lazy where appropriate:
- even if an expression used, it might be constant (pulling in some config variables, etc...)/could be mentioned in a static constant
- it's easy enough to write something like

```
    static SHEET_REF: Lazy<Sheet> = Lazy::new(|| {
        sheet! {
            &:has(> img) {
                backgroundColor: black;
            }
        }
    });
```
- it actually overcomplicates the behaviour of the macro
- you're writing Rust. If you want performance there is some thought to be given. An macro expanding to an expression, being used in an expression context, expanding to something behind a static Lazy is not my first thought of how it behaves.

futursolo replied:
I don't think it's overcomplicating things.

One fundamental point of macros is to provide optimisation by selecting the most optimised implementation without user intervention.

Example: https://doc.rust-lang.org/src/alloc/macros.rs.html#43

In addition, I don't think simplification need to be made at a cost of user experience.

Why do we need to tell user to optimise themselves if this optimisation can be made by procedural macros automatically?

> Instead of doing
> 
> ```rust
> html! {<div class=css!{ color: red; } />}
> ```
> 
> you should be doing
> 
> ```rust 
> static SHEET: Lazy<Sheet> = Lazy::new(|| sheet! { color: red; });
> html! {<div class=Style::new(SHEET.clone()) />}
> ```
> 
> if you want the best performance.

And imagine you have to do that for every component if you want great performance.
(I cannot imagine anyone would be bothered to do this for every stylesheet if they are maintaining a sizable website.)

> you're writing Rust. If you want performance there is some thought to be given. An macro expanding to an expression, being used in an expression context, expanding to something behind a static Lazy is not my first thought of how it behaves.

How would it behave differently to an end user comparing to a not Lazy'ed variant other than it being slower? And I don't think most users would want to know what code a macro expands to.

In addition, this also helps to mitigate inefficiency in generated code (excessive `StringFragment` when not needed, etc.), in which I will address in the future but haven't had time to do so.

When creating a `Sheet`, there're a lot of small allocations, which is not efficient to do (even worse when people are using `wee_alloc`.). Hence it's somewhat critical to have a `Lazy` in the macro.

```
Benchmark	Result
Parse Simple (1,000,000 iterations):	70ms
Macro Simple (1,000,000 iterations):	5ms
Parse Complex (100,000 iterations):	56ms
Macro Complex (100,000 iterations):	1ms
```

Benchmark when `Lazy` is removed.
```
Benchmark	Result
Parse Simple (1,000,000 iterations):	73ms
Macro Simple (1,000,000 iterations):	79ms (Slower than runtime cache?)
Parse Complex (100,000 iterations):	56ms
Macro Complex (100,000 iterations):	44ms
```

WorldSEnder replied:
Those benchmarks are very persuasive :) I'll have to probably inspect the expressions inside `InjectedExpression`s also, since e.g. `${"2em"}` (a literal fragment being injected) should trigger the optimization as well.

WorldSEnder commented on packages/stylist-macros/src/tokentree/output/mod.rs:
```diff
@@ -0,0 +1,44 @@
+use proc_macro2::TokenStream;
+
```
The `Output*` types are there to mirror the structure in `stylist::ast`. Going through them 1-to-1 should give a nice readable way to follow changes to the ast and how each struct there is generated.

futursolo replied:
I see. I guess we can keep the `Output*`, but I still kind of want to move parsing logic of selectors out of it.

I have also noticed a lot of following pattern:

```rust
        let ident_scopes = Ident::new("scopes", Span::mixed_site());
        let Self { contents } = self;

        quote! {
            {
                let #ident_scopes: ::std::vec::Vec::<::stylist::ast::ScopeContent> = ::std::vec![
                    #( #contents, )*
                ];
                ::stylist::ast::Sheet::from(#ident_scopes)
            }
        }
```

over

```rust
quote! { ::stylist::ast::Sheet::from(::std::vec![#( #contents, )*]) }
```

I thought `Span`s are used for when procedural macros are having an error to emit diagnostic information.
Is there other advantage of of casting a span on the generated token if it can never error here?

WorldSEnder replied:
Assigning to `#ident_scopes` here to make the type checker happy. An alternative would be

`quote! { ::stylist::ast::Sheet::from(::std::vec![#( ::stylist::ast::ScopeContent::from(#contents), )*]) }`

Which reminds me to put a `use std::convert::{Into, From};` into the quote.

futursolo commented:
Other than the complexity level of code and some inefficiency in code generation (which I think it's ok to address later), I think this PR looks solid.

I do like this syntax as it is simpler than string literals when it comes to interpolations.

My rationale about reducing code complexity is based on that it may hinder future contribution from other developers if they cannot easily understand the parser internals.
I wish the codebase to stay friendly to a first-time contributor.

Thank you for another big PR.

WorldSEnder commented on packages/stylist-macros/src/tokentree/parsed.rs:
```diff
@@ -0,0 +1,535 @@
+/// outer-attribute: some;
+/// foo-bar: zet;
+/// .nested {
+///     @media print {
```
Just trying to mirror other libraries. It's sometimes very useful to have this kind of nesting in source code. I do not know how to best port this to the other parser.

futursolo replied:
I guess maybe we can get away with this feature in this round. I don't think we can justify only one syntax flavour supports a feature and the other one doesn't.

If we remove it from this round of implementation of then we can also remove `fold_*` (which greatly reduces code complexity) as I think manipulating ast here is probably not the proper way to implement this feature if both sides introduces this syntax.

It's easy enough to write something like (for the moment):
```css
@media screen and (max-width: 500px) {
    .selector { color: red; }
}

@media screen and (max-width: 600px) {
    .selector { color: blue; }
}

@media screen and (max-width: 700px) {
    .selector { color: yellow; }
}
```

WorldSEnder replied:
Okay with me. I had planned to use it, e.g. to write

```css
&:hover {
    border: 1px solid ${color_secondary.main};
    background-color: ${to_hover(color_secondary.main)};
    @media (hover: none) {
        background-color: transparent;
    }
}
```
instead of
```css
&:hover {
    border: 1px solid ${color_secondary.main};
    background-color: ${to_hover(color_secondary.main)};
}
@media (hover: none) {
    &:hover { background-color: transparent; }
}
```

WorldSEnder replied:
Disallowed for now, easy enough to switch back on

WorldSEnder commented on packages/stylist-macros/src/tokentree/css_ident.rs:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    pub fn stringify(&self) -> String {
```
Changed to `to_name_string` (`to_string` gets denied by clippy). `reify` -> `into_token_stream`

WorldSEnder commented:
I hear you. I can easily see that `fold_tokens_impl` has by far the highest complexity and will think of a way to reduce this before marking the PR as ready for merging and a final review. Thanks for the comments :+1:

WorldSEnder commented on examples/benchmarks/src/benchmarks.rs:
```diff
@@ -53,6 +62,36 @@ pub fn bench_parse_complex() -> f64 {
 
+pub fn bench_macro_complex() -> f64 {
+    let start_time = now();
+    for _ in 0..100_000 {
```
Rewritten to a benchmark that uses the cache for 999/1000 iterations, but demonstrates non `Lazy` optimized behavior.

WorldSEnder commented:
I think I've addressed all review comments thus far. Before I unmark this as Draft, I wanted to check if I should do an interactive rebase and try to rewrite the history a bit so it's cleaner. I think it should be possible to split this into:

- Moving the previous macro into subdirectory
- Adding the new implementation
- Adding test cases
- Adding benchmarks

And have 4 separate commits that each should be more easy to review and read in the commit history. Let me know what you think @futursolo or if there's anything else to do before that, thank you for your comments and nice discussion.

futursolo commented:
In the future, it might be better to separate the changeset into multiple PRs for big changes like this.
For this time, I think it's fine to leave the git history as-is.

Thank you.

futursolo commented on packages/stylist-macros/src/inline/output/maybe_static.rs:
```diff
@@ -0,0 +1,153 @@
+    pub fn into_cow_vec_tokens(self, typ: TokenStream) -> MaybeStatic<TokenStream> {
+        let contents = self.value.into_iter();
+        let quoted = quote! {
+            ::std::borrow::Cow::<'_, [#typ]>::Owned(
```
This type casting is redundant, you can simply return a vec here and use `.into()` in other structs.

WorldSEnder replied:
Just wanted to make sure to fix the correct type and not run into ambiguous types here, as the method is used in a few different places. I think it's fine, as in the future, the const-context optimized version below will also return a `Cow` and the compiler shouldn't care. Should I add a comment explaining this?

futursolo replied:
I don't think we need to consider a feature that is not yet available.

We don't know what approach is the `&'static []` going to take.

I think it's not possible without using `Lazy` given how currently AST is structured.

Which I haven't tested the performance of having multiple levels of Lazy.

WorldSEnder replied:
I think it shouldn't be too hard if we expose `const fn` constructors from `stylist_core` (instead of using `into` inside the proc macros). It was going to be my next PR, since I think it can save quite a bit of performance.

futursolo reviewed (changes requested):
I think there're still some trivial things that I didn't cover.

But I think it might be easier to fix them directly after merging this PR than pointing all of them out in the comments.

Again, thank you for the PR.

WorldSEnder commented on packages/stylist-macros/src/inline/css_ident.rs:
```diff
@@ -0,0 +1,119 @@
+        }
+    }
+
+    pub fn to_name_string(&self) -> String {
```
That was intentional: I think a Display implementation should be what appears in errors if verification of attribute names, et al is implementation in the future and differs, at least conceptually, from the one that returns the contents of the string literal used in quote.

futursolo replied:
In that case, you can still use `format!("{}");`. If clippy lints the generated code, you can `#[allow(clippy::useless_format)]`.

See: https://rust-lang.github.io/rust-clippy/stable/index.html

WorldSEnder replied:
I see it like this:
- `Debug`: what we want to see if we debug the macro
- `Display`: what the user sees if the `CssIdent` shows up in an error message
- `to_name_string`: what gets written into the output css

I'm not sure either pair of these three should always result in the same string.

futursolo replied:
> Display: what the user sees if the CssIdent shows up in an error message (currently exactly what to_name_string returns. I now see what you mean, but I don't want to commit to it staying that way)

I cannot say you are wrong, but with this understanding we cannot explain why interpolated values use `Display` value when printed into the stylesheet.

What you are looking for here is probably a Python's `repr(x)` equivalent which returns a representation of value when invoked. Which in rust, it is usually `{:?}` (Debug).

WorldSEnder replied:
> What you are looking for here is probably a Python's `repr(x)` equivalent which returns a representation of value when invoked. Which in rust, it is usually `{:?}` (Debug). Not strictly equivalent, but usually so.

What I try to say is connected to `trycase` (which I have removed for now for simplicity). Consider:
```
background_color: black;
```
Currently not supported, cause I removed the case transformation, but I did already get a question regarding that on discord. In any case, as implemented in this pr:
```
Debug: "CssIdent { parts: [IdentPart::Ident("background_color")] }"
Display: "background_color"
to_name_string: "background_color"
```
with case transformation added
```
Debug: "CssIdent { parts: [IdentPart::Ident("background_color")] }"
Display: "background_color" <-- this should show in most error messages, as it's what the user wrote?
to_name_string: "background-color" <-- this gets written into the css
```

> with this understanding we cannot explain why interpolated values use Display value when printed into a stylesheet.

I'm not sure I understand what interpolated values have to do with `CssIdent`, but in either case I do think that at some point switching to a custom trait instead of reusing `Display` could be beneficial.

futursolo replied:
Hmm, I see.

My rationale is always that we don't have to consider a feature until we actually lands it so our code base is optimised for what we have at the moment not something we don't have at the moment.

But I guess `to_output_string` would be good enough.

So Css* variant is more based on what's in the code, and `Output*` is more based on what's going to be in `stylist::ast`.

WorldSEnder replied:
> So Css* variant is more based on what's in the code, and `Output*` is more based on what's going to be in `stylist::ast`.

Correct, that's what it's supposed to mean.

WorldSEnder commented on packages/stylist-macros/src/inline/output/maybe_static.rs:
```diff
@@ -0,0 +1,153 @@
+// created everytime anew.
+pub struct MaybeStatic<T> {
+    value: T,
+    context: ExpressionContext,
```
I agree that using the `&` might not be ideal, I changed it after previously having a `bool` and using `&&` which can't overloaded. I'm not sure I understand what you mean by
> with a counter on number of interpolated expressions
I thought maybe using `+`? Technically, `ExpressionContext` is some sort of join-semilattice, but there is no supremum `⋁` operator, so I used `&` for that.

futursolo replied:
So instead of Dynamic or Static, we collect simply collect `num_interpolations`.
If num_interpolations > 0, it means it's dynamic.

WorldSEnder replied:
In a clean design, I think this amounts to passing down a structure through `Reify::into_token_stream` that records this count, so `Reify::into_token_stream(self, &mut Recorder) -> TokenStream`, with `args_used: &mut HashSet<String>` being the prototypical `Recorder` used in `literal`.

[This might also solve a problem of duplicated expressions in nested blocks in the future, if the recorder allows to record expressions and emits them up front.]

An unusual way of a writer-monad for me, coming from functional Haskell, but I'll see if I can make it work :+1:

futursolo replied:
`Reify::into_token_stream(self, &mut Recorder) -> TokenStream`

I agree it's the proper way of doing it.

Although, using a `&Recorder` to collect contextual information might be overkill at the moment, as it only collect a `bool`.

> An unusual way of a writer-monad for me, coming from functional Haskell, but I'll see if I can make it work 👍

I do see a lot of extensive functional language usage (like: iterator) which I was wondering why is it not dealt with using a for-loop. :)

WorldSEnder commented on packages/stylist-macros/src/inline/component_value/mod.rs:
```diff
@@ -0,0 +1,189 @@
+}
+
+impl ComponentValue {
+    pub fn to_output_fragments(&self) -> impl '_ + IntoIterator<Item = OutputFragment> {
```
Ah, probably leftover from when it returned a `Box<dyn Iterator>`. Will change.

WorldSEnder commented on packages/stylist-macros/src/inline/output/str_frag.rs:
```diff
@@ -0,0 +1,148 @@
+
+#[derive(Debug)]
+pub enum OutputFragment {
+    Raw(TokenStream),
```
It's important in `fragment_spacing` to know the different kinds of tokens.

futursolo replied:
> It's important in `fragment_spacing` to know the different kinds of tokens.

I see.

I think fragment spacing should be address with ComponentValue. But guess it's OK to keep for now.

WorldSEnder commented on packages/stylist-macros/src/inline/output/str_frag.rs:
```diff
@@ -0,0 +1,148 @@
+            {
+                fn #ident_write_expr<V: ::std::fmt::Display>(v: V) -> ::std::string::String {
+                    use ::std::fmt::Write;
+                    let mut #ident_result = ::std::string::String::new();
```
Though the compilation error will be different. I will change the internals of `#ident_write_expr` to simply do `to_string()`. Like this, the user gets "does not implement ::std::fmt::Display" compared to "method to_string not found on type", which should be much clearer.

futursolo replied:
In that case, you can still use `format!("{}");`. If clippy lints the generated code, you can `#[allow(clippy::useless_format)]`.

See: https://rust-lang.github.io/rust-clippy/stable/index.html

WorldSEnder replied:
> My original rationale is, IIRC, `to_string` is faster and easier to write

See also the newly add `uses-display-impl` test case why just writing `{ #expr }.to_string()` is not sufficient and runs into edge cases.
